theano/tensor/basic.py

```diff
@@ -2312,17 +2315,22 @@ def ones_like(model, dtype=None):
     """
     if dtype is None:
         dtype = model.type.dtype
-    ret = fill(model, constant(1.0, dtype=dtype))
-    return ret
+    ret = constant(1.0, dtype=dtype)
+    if opt and ret.type == model.type:
+        return ret
+    return fill(model, ret)
```

Comment:
We tried and this make many tests fail.

The cause is that when the user build the graph, they suppose model is in the graph. So it cause error like inputs to theano.function aren't used in the graph.